### PR TITLE
v2.0 in the making

### LINKED ISSUE #146: Handle escaping emulation
Emulation escapes

```
$ floss tests/src/decode-to-stack-rep-mov/bin/test-decode-to-stack-rep-mov.exe -d
[...]
DEBUG:Monitor:prehook: 0x401b5b: call 0x00402900
DEBUG:Monitor:apicall: apicall: <viv_utils.emulator_drivers.DebuggerEmulatorDriver object at 0x7fa2e79fcf50> call 0x00402900 4204800 ('int', None, 'cdecl', 'msvcrt.fwrite', [('void *', 'ptr'), ('int', 'arg1'), ('int', 'arg2'), ('int', 'arg3')]) [4211072, 1, 27, 1096544320]
DEBUG:Monitor:prehook: 0x402900: jmp dword [0x004061b4]
DEBUG:Monitor:prehook: 0x415c1000: inc ecx
DEBUG:Monitor:prehook: 0x415c1001: inc ecx
DEBUG:Monitor:prehook: 0x415c1002: inc ecx
DEBUG:Monitor:prehook: 0x415c1003: inc ecx
DEBUG:Monitor:prehook: 0x415c1004: inc ecx
DEBUG:Monitor:prehook: 0x415c1005: inc ecx
DEBUG:Monitor:prehook: 0x415c1006: inc ecx
DEBUG:Monitor:prehook: 0x415c1007: inc ecx
DEBUG:Monitor:prehook: 0x415c1008: inc ecx
[...]
```

### LINKED ISSUE #271: FLOSS stack strings differs from smartidb stack strings
Missing a bunch of stack strings that our current smartidb plugin catches:
Example MD5: 4cd323ecacdc12367389d25413f5923a

### LINKED ISSUE #277: implement memset hook
review sample from rich for testing

### LINKED ISSUE #282: [identification] functions that decode a bunch of strings
Could look for functions that get called only onces towards the beginning of a binary and write to memory. See 1cb673679f37b6a3f482bb59b52423ab.viv (decoding function at 0x100039A4)

### LINKED ISSUE #299: FLOSS fail 47ac27f6812fdfe243c363edb5ce28df
Tom: 47ac27f6812fdfe243c363edb5ce28df has a simple string decode function @ 401290 and FLOSS is not decoding the strings. I think it’s because it’s not using XOR, but SUB.

### LINKED ISSUE #411: consolidate filtering and sanitization
currently, we spread the logic of filtering and sanitizating all over the place, including:

  - utils: https://github.com/fireeye/flare-floss/blob/6670c059d4d2fe714cf732ad4662f0be24f31ffd/floss/utils.py#L90-L118
  - main: https://github.com/fireeye/flare-floss/blob/6670c059d4d2fe714cf732ad4662f0be24f31ffd/floss/main.py#L76-L84
  - main: https://github.com/fireeye/flare-floss/blob/6670c059d4d2fe714cf732ad4662f0be24f31ffd/floss/main.py#L352-L360

and we propose:

  - #406 stdlib strings
  - #362 flirt strings
  - #405 instruction strings

we should figure out how to do this filtering in a few/one place and make it easy to understand.

### LINKED ISSUE #436: emulation fails on 3aa7ee4d67f562933bc998f352b1f319
emulator doesn't terminate, seems to get stuck on
```
DEBUG: Monitor: prehook: 0x70543ed2: rep: movsb
```

### LINKED ISSUE #460: decoding_manager: SegmentationViolation('Bad Memory Access: 0x415c100f')
`floss tests/data/src/decode-substitution-cipher/bin/test-decode-substitution-cipher.exe --no-static-strings --no-stack-strings --no-tight-strings -dd` (uses new TRACE debugging, use `-d` on older code)

```
...
TRACE: floss: 0x401ad7 mov dword [esp + 12],eax
TRACE: floss: 0x401adb call 0x00402878
TRACE: floss: 0x402878 call 0x00402878 ('int', None, 'cdecl', 'msvcrt.fwrite', [('void *', 'ptr'), ('int', 'arg1'), ('int', 'arg2'), ('int', 'arg3')]) [4211120, 1, 27, 1096544335]
TRACE: floss: 0x402878 jmp dword [0x004061b0]
DEBUG: floss.decoding_manager: vivisect encountered an unexpected exception. will continue processing.
Traceback (most recent call last):
  File "Exclusions\flare-floss\floss\decoding_manager.py", line 165, in emulate_function
    driver.runToVa(return_address, max_instruction_count)
  File "viv-utils\viv_utils\emulator_drivers.py", line 338, in runToVa
    self.stepi()
  File "viv-utils\viv_utils\emulator_drivers.py", line 298, in stepi
    return self.step(False)
  File "viv-utils\viv_utils\emulator_drivers.py", line 280, in step
    op = emu.parseOpcode(startpc)
  File "vivisect\vivisect\impemu\emulator.py", line 242, in parseOpcode
    return self.vw.parseOpcode(va, arch=arch)
  File "vivisect\vivisect\__init__.py", line 1131, in parseOpcode
    off, b = self.getByteDef(va)
  File "vivisect\envi\memory.py", line 567, in getByteDef
    raise envi.SegmentationViolation(va)
envi.exc.SegmentationViolation: SegmentationViolation('Bad Memory Access: 0x415c100f')
DEBUG: floss.decoding_manager: Ended emulation at 0x415C100F
...
```

### PR DESCRIPTION
**Pull request by mr-tz**
A monster PR bringing many improvements.

Need to add more tests with various real-world samples to prove everything still works as expected.

I won't wait for a review but will leave it open for comments for a couple of days.

Also:
closes #411
closes #277
closes #460
closes #436
closes #299
closes #282
closes #271
closes #146

### CODE REVIEW

File: floss/const.py [hunk 1]
```diff
@@ -4,3 +4,7 @@
 SUPPORTED_FILE_MAGIC = {b"MZ"}
 DEFAULT_MIN_LENGTH = 4
 MAX_STRING_LENGTH = 2048
+DS_FUNCTION_CTX_THRESHOLD = 15
+DS_FUNCTION_MIN_DECODED_STRINGS = 5
+TS_MAX_INSTR_COUNT = 10000
+TIGHT_FUNCTION_MAX_BLOCKS = 10
```
Comment (williballenthin) [inline floss/const.py:10]: we should document what these mean

File: floss/api_hooks.py [hunk 2]
```diff
@@ -399,6 +386,25 @@ def hook(self, callname, driver, callconv, api, argv):
...
 
 
+class MemsetHook(viv_utils.emulator_drivers.Hook):
+    """
+    Hook and handle calls to memset
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(MemsetHook, self).__init__(*args, **kwargs)
+
```
[the remark this answers is not in the document]
Comment (mr-tz) [reply floss/api_hooks.py]: thanks, removed some others as well, see a4846cd

File: floss/api_hooks.py [hunk 3]
```diff
@@ -399,6 +386,25 @@ def hook(self, callname, driver, callconv, api, argv):
...
+    def hook(self, callname, driver, callconv, api, argv):
+        if callname == "msvcrt.memset":
+            emu = driver
+            ptr, value, num = argv
+            value = bytes([value]) * num
```
[the remark this answers is not in the document]
Comment (mr-tz) [reply floss/api_hooks.py]: 👍  a4846cd

File: floss/const.py (hunk 1 above)
Comment (mr-tz) [reply floss/const.py]: yep! a4846cd

**Comment by r0ny123** [comment]
Hi @mr-tz. The option for printing function offsets which may have some decoding routine, got disabled?

**Comment by mr-tz** [comment]
Do you mean for the function identification phase? You can still see those results using the debugging `-d` output.

**Comment by r0ny123** [comment]
Ah yes. Just ran the it using -d, can see the decoding functions, but can we print them in the `verbose`(-v) mode?

**Comment by mr-tz** [comment]
Would it work if we include a list of them in the final output table?

```
| decoding functions     | 0x41528C, 0x41A014, 0x413BB4, 0x41893C, 0x4128A4, 0x41762C, ... |
```

**Comment by r0ny123** [comment]
Yes! More suitable.

**Pull request merged**